source/s3_auto_ranged_put.c

```diff
@@ -1102,61 +1260,71 @@ static void s_s3_auto_ranged_put_request_finished(
-                        "id=%p Could not find ETag header for request %p",
-                        (void *)meta_request,
-                        (void *)request);
+                    if (aws_http_headers_get(
+                            request->send_data.response_headers, g_etag_header_name, &etag_within_quotes) !=
+                        AWS_OP_SUCCESS) {
+                        AWS_LOGF_ERROR(
+                            AWS_LS_S3_META_REQUEST,
+                            "id=%p Could not find ETag header for request %p",
+                            (void *)meta_request,
+                            (void *)request);
 
-                    error_code = AWS_ERROR_S3_MISSING_ETAG;
-                } else {
-                    /* The ETag value arrives in quotes, but we don't want it in quotes when we send it back up
-                     * later, so just get rid of the quotes now. */
-                    etag = aws_strip_quotes(meta_request->allocator, etag_within_quotes);
+                        error_code = AWS_ERROR_S3_MISSING_ETAG;
+                    } else {
+                        /* The ETag value arrives in quotes, but we don't want it in quotes when we send it back up
+                         * later, so just get rid of the quotes now. */
+                        etag = aws_strip_quotes(meta_request->allocator, etag_within_quotes);
+                    }
+                }
+                if (error_code == AWS_ERROR_SUCCESS && meta_request->progress_callback != NULL) {
+                    struct aws_s3_meta_request_progress progress = {
+                        .bytes_transferred = meta_request->part_size,
+                        .content_length = auto_ranged_put->content_length,
+                    };
+                    meta_request->progress_callback(meta_request, &progress, meta_request->user_data);
                 }
             }
-            if (error_code == AWS_ERROR_SUCCESS && meta_request->progress_callback != NULL) {
-                struct aws_s3_meta_request_progress progress = {
-                    .bytes_transferred = meta_request->part_size,
-                    .content_length = auto_ranged_put->content_length,
-                };
-                meta_request->progress_callback(meta_request, &progress, meta_request->user_data);
-            }
+
             /* BEGIN CRITICAL SECTION */
             {
                 aws_s3_meta_request_lock_synced_data(meta_request);
 
                 ++auto_ranged_put->synced_data.num_parts_completed;
 
+                if (request_is_noop) {
+                    ++auto_ranged_put->synced_data.num_parts_noop;
+                }
+
                 AWS_LOGF_DEBUG(
                     AWS_LS_S3_META_REQUEST,
                     "id=%p: %d out of %d parts have completed.",
                     (void *)meta_request,
                     auto_ranged_put->synced_data.num_parts_completed,
                     auto_ranged_put->synced_data.total_num_parts);
```

Comment:
We only increase the `total_num_parts` when we successfully read data, which means if the request is_noop, the total num parts will NOT count it. But the `num_parts_completed` will still be increased for noop requests. That will be confusing.

Reply:
fair call out. moved total num parts to always increment